From 685227db25d061d837bbd0939be30006336b538e Mon Sep 17 00:00:00 2001
From: Haihua Hu <jared.hu@nxp.com>
Date: Wed, 14 Dec 2016 15:17:06 +0800
Subject: [PATCH 3/3] Support i.MX 7ULP in imx-gst1.0-plugin

1. Add CHIP_CODE CC_MX7ULP
2. Add platform MX7ULP
3. Add USE_FB_DISPLAY to configue overlaysink display
4. Add protection for xoverlay->disp to fix seg fault
   when call XCloseDisplay

Signed-off-by: Haihua Hu <jared.hu@nxp.com>
---
 Makefile.am                                  | 11 +++++
 configure.ac                                 | 11 +++--
 libs/gstimxcommon.c                          |  2 +
 libs/gstimxcommon.h                          |  1 +
 libs/video-overlay/gstimxxoverlay.c          |  3 +-
 plugins/overlay_sink/Makefile.am             | 22 ++++-----
 plugins/overlay_sink/displaysfb.c            | 73 +++++++++++++++++-----------
 plugins/overlay_sink/imx_7ulp_display_config |  4 ++
 8 files changed, 82 insertions(+), 45 deletions(-)
 mode change 100644 => 100755 libs/gstimxcommon.c
 create mode 100644 plugins/overlay_sink/imx_7ulp_display_config

diff --git a/Makefile.am b/Makefile.am
index f311e8b..3189284 100755
--- a/Makefile.am
+++ b/Makefile.am
@@ -128,6 +128,17 @@ VPUWRAPDIRS =
 
 endif
 
+if PLATFORM_IS_MX7ULP
+V4LSINKDIRS =
+
+if USE_OVERLAY_SINK
+OVERLAYSINKDIRS = plugins/overlay_sink
+endif
+
+VPUWRAPDIRS =
+
+endif
+
 SUBDIRS =  $(LIBSDIRS) $(BASEDIRS) $(V4LSINKDIRS) $(OVERLAYSINKDIRS) $(VPUWRAPDIRS) $(WMA8ENC_DIR) $(MP3ENC_DIR) $(TOOLDIRS)
 
 
diff --git a/configure.ac b/configure.ac
index c325b17..2befa02 100755
--- a/configure.ac
+++ b/configure.ac
@@ -127,11 +127,11 @@ if test "x$HAVE_PKGCONFIG" = "xno"; then
 fi
 
 dnl check target type
-if  test "$PLATFORM" = "MX6" || test "$PLATFORM" = "MX6SL" || test "$PLATFORM" = "MX6SX" || test "$PLATFORM" = "MX7D" || test "$PLATFORM" = "MX6UL" || test "$PLATFORM" = "MX6QP" || test "$PLATFORM" = "MX8" ;
+if  test "$PLATFORM" = "MX6" || test "$PLATFORM" = "MX6SL" || test "$PLATFORM" = "MX6SX" || test "$PLATFORM" = "MX7D" || test "$PLATFORM" = "MX6UL" || test "$PLATFORM" = "MX6QP" || test "$PLATFORM" = "MX8" || test "$PLATFORM" = "MX7ULP";
 then
   AC_MSG_NOTICE(build for $PLATFORM)
 else
-  AC_MSG_ERROR(No target platform specified! Use ./configure PLATFORM=(MX6/MX6QP/MX6SL/MX6SX/MX6UL/MX7D) to retry)
+  AC_MSG_ERROR(No target platform specified! Use ./configure PLATFORM=(MX6/MX6QP/MX6SL/MX6SX/MX6UL/MX7D/MX8/MX7ULP) to retry)
 fi
 
 AC_SUBST(PLATFORM)
@@ -332,9 +332,12 @@ AM_CONDITIONAL(PLATFORM_IS_MX6SX, test "x$PLATFORM" = "xMX6SX")
 AM_CONDITIONAL(PLATFORM_IS_MX6UL, test "x$PLATFORM" = "xMX6UL")
 AM_CONDITIONAL(PLATFORM_IS_MX6QP, test "x$PLATFORM" = "xMX6QP")
 AM_CONDITIONAL(PLATFORM_IS_MX7D, test "x$PLATFORM" = "xMX7D")
-AM_CONDITIONAL(PLATFORM_IS_ARCH32, test "x$PLATFORM" = "xMX7D" || test "x$PLATFORM" = "xMX6" || test "x$PLATFORM" = "xMX6" || test "x$PLATFORM" = "xMX6SL" || test "x$PLATFORM" = "xMX6SX" || test "x$PLATFORM" = "xMX6UL" || test "x$PLATFORM" = "xMX6QP")
+AM_CONDITIONAL(PLATFORM_IS_ARCH32, test "x$PLATFORM" = "xMX7D" || test "x$PLATFORM" = "xMX6" || test "x$PLATFORM" = "xMX6" || test "x$PLATFORM" = "xMX6SL" || test "x$PLATFORM" = "xMX6SX" || test "x$PLATFORM" = "xMX6UL" || test "x$PLATFORM" = "xMX6QP" || test "x$PLATFORM" = "xMX7ULP")
+AM_CONDITIONAL(PLATFORM_IS_MX7ULP, test "x$PLATFORM" = "xMX7ULP")
 AM_CONDITIONAL(PLATFORM_IS_MX8, test "x$PLATFORM" = "xMX8")
 
+AM_CONDITIONAL(USE_FB_DISPLAY, test "x$PLATFORM" = "xMX8" || test "x$PLATFORM" = "xMX7ULP")
+
 if test "x$PLATFORM" = "xMX6" || test "x$PLATFORM" = "xMX6QP"; then
 	 HAVE_DEVICE_IPU=yes
 	 HAVE_DEVICE_G2D=yes
@@ -349,7 +352,7 @@ if test "x$PLATFORM" = "xMX6UL" || test "x$PLATFORM" = "xMX7D"; then
 	 HAVE_DEVICE_PXP=yes
 fi
 
-if test "x$PLATFORM" = "xMX8" || test "x$PLATFORM" = "xMX8"; then
+if test "x$PLATFORM" = "xMX8" || test "x$PLATFORM" = "xMX7ULP"; then
 	 HAVE_DEVICE_G2D=yes
 fi
 
diff --git a/libs/gstimxcommon.c b/libs/gstimxcommon.c
old mode 100644
new mode 100755
index 7742e08..6765754
--- a/libs/gstimxcommon.c
+++ b/libs/gstimxcommon.c
@@ -114,6 +114,7 @@ static SOC_INFO soc_info[] = {
   {CC_MX6UL, "i.MX6UL"},
   {CC_MX6UL, "i.MX6ULL"},
   {CC_MX7D, "i.MX7D"},
+  {CC_MX7ULP, "i.MX7ULP"},
   {CC_MX8, "i.MX8DV"},
 };
 
@@ -193,6 +194,7 @@ static IMXV4l2FeatureMap g_imxv4l2feature_maps[] = {
   {CC_MX6SX, TRUE, TRUE, FALSE, TRUE, FALSE, FALSE},
   {CC_MX6UL, FALSE, FALSE, FALSE, TRUE, FALSE, FALSE},
   {CC_MX7D, FALSE, FALSE, FALSE, TRUE, FALSE, FALSE},
+  {CC_MX7ULP, TRUE, TRUE, FALSE, FALSE, FALSE, FALSE},
   {CC_MX8, TRUE, TRUE, FALSE, FALSE, FALSE, TRUE},
 };
 
diff --git a/libs/gstimxcommon.h b/libs/gstimxcommon.h
index 0fae8f0..51005dd 100755
--- a/libs/gstimxcommon.h
+++ b/libs/gstimxcommon.h
@@ -62,6 +62,7 @@ typedef enum
   CC_MX6SX = CHIPCODE ('M', 'X', '6', '2'),
   CC_MX6UL = CHIPCODE ('M', 'X', '6', '3'),
   CC_MX7D = CHIPCODE ('M', 'X', '7', 'D'),
+  CC_MX7ULP = CHIPCODE ('M', 'X', '7', 'U'),
   CC_MX8 = CHIPCODE ('M', 'X', '8', '0'),
   CC_UNKN = CHIPCODE ('U', 'N', 'K', 'N'),
 
diff --git a/libs/video-overlay/gstimxxoverlay.c b/libs/video-overlay/gstimxxoverlay.c
index f8c1975..72126c0 100755
--- a/libs/video-overlay/gstimxxoverlay.c
+++ b/libs/video-overlay/gstimxxoverlay.c
@@ -413,7 +413,8 @@ gst_x_video_overlay_deinit(ImxVideoOverlay * imxxoverlay)
   if (imxxoverlay->video_win)
     XSelectInput (xoverlay->disp, imxxoverlay->video_win, NoEventMask);
   gst_x_video_overlay_destroy_window(imxxoverlay);
-  XCloseDisplay (xoverlay->disp);
+  if (xoverlay->disp)
+    XCloseDisplay (xoverlay->disp);
   xoverlay->disp = NULL;
   g_mutex_clear (&xoverlay->mutex);
   g_free(xoverlay);
diff --git a/plugins/overlay_sink/Makefile.am b/plugins/overlay_sink/Makefile.am
index 32e5698..57a3235 100755
--- a/plugins/overlay_sink/Makefile.am
+++ b/plugins/overlay_sink/Makefile.am
@@ -1,18 +1,9 @@
 plugin_LTLIBRARIES = libgstoverlaysinkplugins.la
 
-if PLATFORM_IS_MX8
-libgstoverlaysinkplugins_la_SOURCES = gstosink.c \
-              gstosinkallocator.c \
-				      osink_object.c \
-				      compositor.c \
-				      displaysfb.c
-else
 libgstoverlaysinkplugins_la_SOURCES = gstosink.c \
               gstosinkallocator.c \
 				      osink_object.c \
-				      compositor.c \
-				      displaysv4l2.c
-endif
+				      compositor.c
 
 libgstoverlaysinkplugins_la_CFLAGS = $(GST_PLUGINS_BASE_CFLAGS) \
 				 $(GST_BASE_CFLAGS) \
@@ -26,6 +17,12 @@ libgstoverlaysinkplugins_la_CFLAGS = $(GST_PLUGINS_BASE_CFLAGS) \
          -I$(top_srcdir)/ext-includes \
          -I$(top_srcdir)/libs/video-overlay
 
+if USE_FB_DISPLAY
+libgstoverlaysinkplugins_la_SOURCES += displaysfb.c
+else
+libgstoverlaysinkplugins_la_SOURCES += displaysv4l2.c
+endif
+
 if PLATFORM_IS_MX8
 libgstoverlaysinkplugins_la_CFLAGS += -DUSE_FB_API
 endif
@@ -50,6 +47,7 @@ noinst_HEADERS = osink_common.h gstosink.h gstosinkallocator.h osink_object.h co
 registry_file1 = imx_6q_display_config
 registry_file2 = imx_6sx_display_config
 registry_file3 = imx_8dv_display_config
+registry_file4 = imx_7ulp_display_config
 
-data_DATA = $(registry_file1) $(registry_file2) $(registry_file3)
-EXTRA_DIST = $(registry_file1) $(registry_file2) $(registry_file3)
+data_DATA = $(registry_file1) $(registry_file2) $(registry_file3) $(registry_file4)
+EXTRA_DIST = $(registry_file1) $(registry_file2) $(registry_file3) $(registry_file4)
diff --git a/plugins/overlay_sink/displaysfb.c b/plugins/overlay_sink/displaysfb.c
index 9409d7e..5257d07 100755
--- a/plugins/overlay_sink/displaysfb.c
+++ b/plugins/overlay_sink/displaysfb.c
@@ -30,6 +30,7 @@
 #include <linux/mxcfb.h>
 #endif
 
+#include "gstimxcommon.h"
 #include "displays.h"
 #include "gstsutils.h"
 #include "gstimxv4l2.h"
@@ -181,11 +182,19 @@ static void set_display_color_key (gchar *device, gboolean enable, gint color_ke
 
 gint scan_displays(gpointer **phandle, gint *pcount)
 {
-#ifdef USE_FB_API
-  GstsutilsEntry *entry =  gstsutils_init_entry ("/usr/share/imx_8dv_display_config");
-#else
-  GstsutilsEntry *entry =  gstsutils_init_entry ("/usr/share/imx_displays_config");
-#endif
+  GstsutilsEntry *entry = NULL;
+  CHIP_CODE chipcode = imx_chip_code();
+  switch (chipcode) {
+    case CC_MX8:
+      entry = gstsutils_init_entry ("/usr/share/imx_8dv_display_config");
+    break;
+    case CC_MX7ULP:
+      entry = gstsutils_init_entry ("/usr/share/imx_7ulp_display_config");
+    break;
+    default:
+    break;
+  }
+
   gint group_count;
   gint i;
   gint count = 0;
@@ -349,29 +358,33 @@ gint init_display (gpointer display)
     return -1;
   }
 
-  fb_var.xoffset = 0;
-  fb_var.xres = hdisplay->w;
-  fb_var.xres_virtual = hdisplay->w;
-  fb_var.yoffset = 0;
-  fb_var.yres = hdisplay->h;
-  fb_var.yres_virtual = hdisplay->h * DISPLAY_NUM_BUFFERS;
-  fb_var.activate |= FB_ACTIVATE_FORCE;
-  fb_var.nonstd = hdisplay->fmt;
-
-  if (hdisplay->fmt == GST_MAKE_FOURCC('R', 'G', 'B', 'x')
-      || hdisplay->fmt == GST_MAKE_FOURCC('B', 'G', 'R', 'x')
-      || hdisplay->fmt == GST_MAKE_FOURCC('B', 'G', 'R', 'A')
-      || hdisplay->fmt == GST_MAKE_FOURCC('A', 'R', 'G', 'B')) {
-    fb_var.bits_per_pixel = 32;
-  } else if (hdisplay->fmt == GST_MAKE_FOURCC('R', 'G', 'B', 'P')) {
-    fb_var.bits_per_pixel = 16;
-  } else {
-    GST_ERROR ("Unsupported display format, check the display config file.");
-    return -1;
-  }
+  /*FIXME: 7ULP don't support seting fb temporary */
+  CHIP_CODE chipcode = imx_chip_code();
+  if (chipcode != CC_MX7ULP) {
+    fb_var.xoffset = 0;
+    fb_var.xres = hdisplay->w;
+    fb_var.xres_virtual = hdisplay->w;
+    fb_var.yoffset = 0;
+    fb_var.yres = hdisplay->h;
+    fb_var.yres_virtual = hdisplay->h * DISPLAY_NUM_BUFFERS;
+    fb_var.activate |= FB_ACTIVATE_FORCE;
+    fb_var.nonstd = hdisplay->fmt;
+
+    if (hdisplay->fmt == GST_MAKE_FOURCC('R', 'G', 'B', 'x')
+        || hdisplay->fmt == GST_MAKE_FOURCC('B', 'G', 'R', 'x')
+        || hdisplay->fmt == GST_MAKE_FOURCC('B', 'G', 'R', 'A')
+        || hdisplay->fmt == GST_MAKE_FOURCC('A', 'R', 'G', 'B')) {
+      fb_var.bits_per_pixel = 32;
+    } else if (hdisplay->fmt == GST_MAKE_FOURCC('R', 'G', 'B', 'P')) {
+      fb_var.bits_per_pixel = 16;
+    } else {
+      GST_ERROR ("Unsupported display format, check the display config file.");
+      return -1;
+    }
 
-  if (ioctl(hdisplay->fd, FBIOPUT_VSCREENINFO, &fb_var) < 0) {
-    return -1;
+    if (ioctl(hdisplay->fd, FBIOPUT_VSCREENINFO, &fb_var) < 0) {
+      return -1;
+    }
   }
 
   struct fb_fix_screeninfo fb_fix;
@@ -413,7 +426,11 @@ void deinit_display (gpointer display)
   }
 
   if (hdisplay->fd) {
-    ioctl(hdisplay->fd, FBIOBLANK, FB_BLANK_NORMAL);
+    /*FIXME: 7ULP will hang on FBIOBLANK, drivers issue */
+    CHIP_CODE chipcode = imx_chip_code();
+    if (chipcode != CC_MX7ULP) {
+      ioctl(hdisplay->fd, FBIOBLANK, FB_BLANK_NORMAL);
+    }
     munmap(hdisplay->vaddr, hdisplay->fb_size);
     close (hdisplay->fd);
     hdisplay->fd = 0;
diff --git a/plugins/overlay_sink/imx_7ulp_display_config b/plugins/overlay_sink/imx_7ulp_display_config
new file mode 100644
index 0000000..938c110
--- /dev/null
+++ b/plugins/overlay_sink/imx_7ulp_display_config
@@ -0,0 +1,4 @@
+[master]
+device = /dev/fb0
+bg_device = /dev/fb0
+fmt = BGRA
-- 
1.9.1

